File: arduinobot_ws/src/arduinobot_cpp_examples/src/simple_service_client.cpp
/**
 * @file simple_service_client.cpp
 * @brief ROS 2 example: minimal asynchronous service client node.
 *
 * Demonstrates how to create a ROS 2 service client using rclcpp and send an
 * asynchronous request. The node connects to the `add_two_ints` service,
 * waits for it to become available, sends a request with two integers provided
 * as command-line arguments, and prints the response.
 *
 * The service uses the custom message type defined in @c arduinobot_msgs:
 *   - **Request**: two signed integers @c a and @c b (from argv).
 *   - **Response**: their @c sum.
 *
 * @par How to run
 * @code{.sh}
 * # Start the server first (in another terminal)
 * ros2 run arduinobot_cpp_examples simple_service_server
 *
 * # Then run the client with two integer arguments
 * ros2 run arduinobot_cpp_examples simple_service_client 3 7
 * @endcode
 *
 * @author Andres Aleman
 * @date 2024
 */

#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "arduinobot_msgs/srv/add_two_ints.hpp"
    
using namespace std::chrono_literals;
using std::placeholders::_1;

/**
 * @class SimpleServiceClient
 * @brief ROS 2 node that asynchronously calls the `add_two_ints` service.
 *
 * Inherits from @c rclcpp::Node. On construction the node:
 *   1. Creates a service client for `add_two_ints`.
 *   2. Waits in a polling loop (1 s intervals) until the service is available.
 *   3. Sends an asynchronous request with the two integer operands.
 *   4. Processes the response in @ref responseCallback.
 *
 * This example illustrates:
 *   - Creating a typed client with `rclcpp::Node::create_client()`.
 *   - Waiting for a service with `wait_for_service()`.
 *   - Sending asynchronous requests with `async_send_request()`.
 *   - Handling the result in a future-based callback.
 */
class SimpleServiceClient : public rclcpp::Node 
{
public:
    /**
     * @brief Construct a new SimpleServiceClient node and immediately send a request.
     *
     * Blocks (within the constructor) until the `add_two_ints` service becomes
     * available, then fires off the asynchronous request. The response will be
     * delivered to @ref responseCallback.
     *
     * @param a First integer operand (maps to `Request::a`).
     * @param b Second integer operand (maps to `Request::b`).
     */
    SimpleServiceClient(int a, int b) : Node("simple_service_client") 
    {     
        client_ = create_client<arduinobot_msgs::srv::AddTwoInts>("add_two_ints");
        auto request = std::make_shared<arduinobot_msgs::srv::AddTwoInts::Request>();
        request ->a = a;
        request->b = b;

        while (!client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) 
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
                return;
            }
            RCLCPP_WARN(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
        }

        auto result = client_->async_send_request(request, std::bind(&SimpleServiceClient::responseCallback, this, _1));    
    }
    
private:
    /// Async service client handle for the `add_two_ints` service.
    rclcpp::Client<arduinobot_msgs::srv::AddTwoInts>::SharedPtr client_;

    /**
     * @brief Response callback — logs the service result.
     *
     * Called automatically by the ROS 2 executor when the shared future
     * returned by `async_send_request()` becomes ready.
     *
     * @param future Shared future holding the server's response. When valid,
     *               `future.get()->sum` contains the addition result.
     */
    void responseCallback(rclcpp::Client<arduinobot_msgs::srv::AddTwoInts>::SharedFuture future)
    {
        if(future.valid())
        {
            RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "Service Response " << future.get()->sum);
        }
        else
        {
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Service Failure");
        }
    }

};

/**
 * @brief Program entry point.
 *
 * Expects exactly two additional command-line arguments (the two integers to
 * add). Logs an error and returns if the argument count is wrong.
 * Otherwise initialises the ROS 2 runtime, creates @ref SimpleServiceClient,
 * spins the node until shutdown, and cleans up.
 *
 * @param argc Number of command-line arguments (must be 3: program + A + B).
 * @param argv Array of command-line argument strings; `argv[1]` = A, `argv[2]` = B.
 * @return int Returns 0 on clean exit.
 */
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    if(argc != 3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Wrong number of arguments! Usage: simple_service_client A B");
    }

    auto node = std::make_shared<SimpleServiceClient>(atoi(argv[1]), atoi(argv[2]));
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: arduinobot_ws/src/arduinobot_py_examples/arduinobot_py_examples/simple_service_client.py
"""
Simple service client node for the arduinobot_py_examples package.

Provides :class:`SimpleServiceClient`, a ROS 2 node that calls the
``add_two_ints`` service using the :class:`arduinobot_msgs.srv.AddTwoInts`
interface. The two operands are supplied as command-line arguments.
Intended as a teaching example for ROS 2 service client patterns.
"""

import sys
import rclpy
from rclpy.node import Node
from arduinobot_msgs.srv import AddTwoInts


class SimpleServiceClient(Node):
    """A ROS 2 node that asynchronously calls the ``add_two_ints`` service.

    Waits for the service to become available, sends a single request with
    the provided operands, and logs the response via a done callback.

    **Services Used:**
        - ``add_two_ints`` (:class:`arduinobot_msgs.srv.AddTwoInts`): Requests
          the sum of two integers.

    :param a: First integer operand.
    :type a: int
    :param b: Second integer operand.
    :type b: int
    """

    def __init__(self, a, b):
        """Initialize the service client node and dispatch a request.

        Blocks in a polling loop until the ``add_two_ints`` service is
        available, then dispatches an asynchronous call with *a* and *b*
        and registers :meth:`responseCallback` as the completion handler.

        :param a: First integer operand to send in the request.
        :type a: int
        :param b: Second integer operand to send in the request.
        :type b: int
        """
        super().__init__("simple_service_client")
        self.client_ = self.create_client(AddTwoInts, "add_two_ints")

        while not self.client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("service not available, waiting again...")

        self.req_ = AddTwoInts.Request()
        self.req_.a = a
        self.req_.b = b
        self.future_ = self.client_.call_async(self.req_)
        self.future_.add_done_callback(self.responseCallback)

    def responseCallback(self, future):
        """Log the result returned by the ``add_two_ints`` service.

        :param future: Completed future carrying the service response.
        :type future: rclpy.task.Future
        """
        self.get_logger().info('Service Response %d' % future.result().sum)


def main(args=None):
    """Parse command-line arguments and spin up :class:`SimpleServiceClient`.

    Expects exactly two positional arguments after the executable name:
    the integer operands *A* and *B*.

    :returns: ``-1`` if the argument count is wrong, ``None`` otherwise.
    """
    rclpy.init(args=args)

    if len(sys.argv) != 3:
        print("Wrong number of arguments! Usage: Simple_service client A B")
        return -1

    node = SimpleServiceClient(int(sys.argv[1]), int(sys.argv[2]))
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
